### PR TITLE
Simplify error notices for broken connections

### PR DESCRIPTION
**Pull request by leogermani**
#16194 introduced the ability for Jetpack to identify, confirm and handle XML-RPC errors. It also introduced admin notices to display such errors to users. 

Up until now we were only displaying errors for Blog token issues. This PR adds error messages for user tokens errors too.

### Refactor

This PR also simplifies the code that was displaying the errors. At first, there was this idea of having an abstraction that would allow us to treat each error code differently, having a class to handle each one of them.

However, this abstraction was in place to handle theoretical situations while, in practice, we were simply displaying the same generic error message in all case. Therefore I decided to get rid of the abstraction for now and keep it simple. When needed, it's easy to add a per error code handling there.

**Note:** I think it's completely safe to just remove the "blog token error handling" class and all the hooks introduced in 8.7 because they have not been used anywhere or by any other plugin using the connection package.

#### Changes proposed in this Pull Request:

* Adds user feedback for XML-RPC user token related errors

#### Jetpack product discussion

p1HpG7-9q9-p2

#### Does this pull request change what data or activity we track or use?

No

#### Testing instructions:

* Start with a connected Jetpack site and the Jetpack Debug helper plugin activated
* Go to Broken Token and click "Set invalid user token (current user)"
* Go to XML-RPC Errors.
* Click the link to visit the Jetpack debugger
* Back to XML-RPC error, confirm you have an error under "Verified error"
* Visit any Dashboard page and confirm you see the error in admin notices
* Visit the Jetpack Dashboard and confirm you see the error at the top of the page along with a CTA to restore the connection

#### Proposed changelog entry for your changes:

* Adds user feedback for XML-RPC user token related errors

### CODE REVIEW

**Comment by leogermani** [comment]
Thanks @fgiannar 

> Tested this and can verify that I am able to see a `no_valid_token` error in XML-RPC errors, but it appears under **Current Unverified Errors** section instead.

Actually it appears on both Unverified and Verified sections. If you saw the error notices, I trust it was there.

> The error notice and corresponding CTA appeared as expected.

Good

> Finally, a small code related suggestion: Do you believe it would make sense to add an error-handling interface, so that all error handler classes implement a contract for the `admin_notice` method and possibly a message property related getter/setter?

Yes, it definitely makes sense to have an interface. It would be a very generic interface though, because, in theory, each error handler can perform very different actions. Not all of them would necessarily display error on admin notice.

I didn't create it yet because I'm avoiding premature abstraction... While in theory we can have many different types of error handlers, in practice, right now, we only have 2 that are actually identical... they could be the same. They only trigger different actions and filters to allow third party users (also theoretical) to act on these errors. So I'm actually tending to ask if any abstraction there is worth at this point, and if we should just have one simple method or class that adds the error to the admin notice hook. What do you think?

(I'm actually waiting the partial reconnect PR to be merge to make Jetpack plugin use these hooks and customize the error messages, with a button to take the user straight to the reconnect flow.)

**Comment by fgiannar** [comment]
> Actually it appears on both Unverified and Verified sections. If you saw the error notices, I trust it was there.

The only thing that troubles me here is that while testing, I only managed to see it under the Unverified section, which could be inconsistent with the fact that it is displayed as an error to the users (I was under the impression that we only show Verified)

> So I'm actually tending to ask if any abstraction there is worth at this point, and if we should just have one simple method or class that adds the error to the admin notice hook. What do you think?

Having a simple method or class that adds the error to the admin notice hook sounds reasonable.

File: packages/connection/src/class-error-handler.php [hunk 1]
```diff
@@ -580,4 +589,74 @@ public function verify_xml_rpc_error( \WP_REST_Request $request ) {
...
+	 * @since 8.9.0
+	 *
+	 * @param array $errors The array of errors. See Automattic\Jetpack\Connection\Error_Handler for details on the array structure.
+	 * @return array
+	 */
+	public function jetpack_react_dashboard_error( $errors ) {
+
+		$errors[] = array(
+			'code'    => 'invalid_blog_token',
+			'message' => __( 'Your connection with WordPress.com seems to be broken. If you\'re experiencing issues, please try reconnecting.', 'jetpack' ),
```
Comment (leogermani) [inline packages/connection/src/class-error-handler.php:656]: The duplicated error message here is intentional.

Leaving it separated from the filter in the method above allows Jetpack plugin to have full control over it's Dashboard. If a plugin filters the message and modifies or disables it, it won't affect the Jetpack plugin.

**Comment by leogermani** [comment]
@fgiannar I've refactored the code and kept it simpler. We don't need all that abstraction right now. I updated the PR description with some additional info.

I also added a page in the documentation explaining how to customize or disable the error messages :)

File: packages/connection/src/class-error-handler.php [hunk 2]
```diff
@@ -580,4 +589,74 @@ public function verify_xml_rpc_error( \WP_REST_Request $request ) {
...
+	 *
+	 * @since 8.9.0
+	 *
+	 * @param array $errors The array of errors. See Automattic\Jetpack\Connection\Error_Handler for details on the array structure.
+	 * @return array
+	 */
+	public function jetpack_react_dashboard_error( $errors ) {
+
+		$errors[] = array(
+			'code'    => 'invalid_blog_token',
```
[the remark this answers is not in the document]
Comment (leogermani) [reply packages/connection/src/class-error-handler.php:655]: Yeah, this is wrong! Nice catch

**Comment by leogermani** [comment]
> @leogermani
> Tested this using the instructions and noticed the following weird behaviour:
> The first time I invalidate the user token, I can see the XML-RPC error and also the error notice in the Dashboard.
> After restoring my connection, if I try to repeat the above process, I don't see the error at all.
> The same happens when I remove all user related tokens and set up the connection from the admin UI.
> I reproduced this both in my local env and a JN site.
> Could you check and let me know if you can reproduce this as well?

This is expected because we have the gates in place making sure we just run all the [verification process](https://github.com/Automattic/jetpack/pull/16194) once in a while to avoid overflows.

On the plugin side you can easily disable this by setting `define( 'JETPACK_DEV_DEBUG', true );`.

On wpcom side it's more tricky, but here I explain how to do it: D47732-code (you'd have to sandbox `public-api.wordpress.com`.

You can also only disable the gate on the plugin. The error will appear as Unverified. You can then click the "Verify error (via api)" button and manually verify it.

**Comment by leogermani** [comment]
@kraftbj @jeherve This PR is ready to merge but wanted to get your ok on something.

On this PR I'm removing some classes and filters added in 8.7 that I'm sure no one is using (considering woocommerce payments as our only client right now).

Since we had backward compatibility issues before, and this should most likely trigger our checks for removing methods and filters, just wanted to give you a heads up that this should look clean and safe.

**Comment by leogermani** [comment]
@kraftbj @jeherve new ping. I think this might have gone out of the radar

**Comment by jeherve** [comment]
> I'm removing some classes and filters added in 8.7 that I'm sure no one is using (considering woocommerce payments as our only client right now).

Yeah, this should be fine. 👍

**Review by jeherve — APPROVED** [review]

**Pull request merged**